fix(ORDataObject): refactor for consistent initialization and id type

kojiromike opened the pull request:
Fixes #8491

#### Short description of what this resolves:

Makes all the subclasses of `ORDataObject` initialize consistently and have the same type of `$id`.

#### Changes proposed in this pull request:

1. Make `ORDataObject` an abstract class.
2. Implement `get_id` and `set_id` consistently in `ORDataObject` and remove subclass implementations.
3. `set_id` is never fluent.
4. Enable implementations to supply `$_table` and optionally `$_prefix` as class properties instead of setting them in the constructor.
5. Make a consistent default constructor such that
  1. Default value initialization (`init`) and database population (`populate`) can happen for most classes without overriding the constructor.
  2. It's clear when something does need to override the constructor.
6. `ORDataObject`'s own call to `get_id` can reference a function that exists.
7. All subclasses call the constructor, explicitly or implicitly.

#### Does your code include anything generated by an AI Engine? No

kojiromike committed:
fix(ORDataObject): refactor for consistent initialization and id type

adunsulag commented:
@kojiromike did you want this to be a draft?  I noticed a lot of the tests are failing with these changes, wanted to check to see if you wanted us to review this or not.

bradymiller commented:
hi @kojiromike , I'll check this out over weekend if nobody else gets to it by then.

adunsulag committed:
Fix Form Vitals set_id function

The dynamic nature of the set function that was refactored from the
Smarty templates messes up the set_id.  We'll need to look at some of
the other classes and see what is going on.

adunsulag commented:
I pulled this down to play with it from the GUI side, it looks like C_FormVitals::populate_object() was failing.  Sadly, due to our lack of sufficient tests we're going to need to do some tests on the other forms,  if this one was broken there are likely to be others :(

kojiromike commented on interface/forms/vitals/C_FormVitals.class.php:
```diff
@@ -450,9 +450,15 @@ public function populate_object(&$obj)
         foreach ($_POST as $varname => $var) {
             $varname = preg_replace("/[^A-Za-z0-9_]/", "", $varname);
             $func = "set_" . $varname;
+            if ($func == "set_id") {
+                if (!empty($var)) {
+                    $var = (int)$var; // make sure we have an integer
+                } else {
+                    $var = null; // if we don't have an id then we set it to null
+                }
+            }
```
I take it 0 is not a valid value for the id?

kojiromike replied:
It just seems it might confuse readers, because empty(0) is true, but 0 is a valid integer.

kojiromike commented:
> I pulled this down to play with it from the GUI side, it looks like C_FormVitals::populate_object() was failing. Sadly, due to our lack of sufficient tests we're going to need to do some tests on the other forms, if this one was broken there are likely to be others :(

Understood. I don't know when I'll get to it, but I'll see what I can do to write tests on the old version of the code, before taking this change through.
